ember-core/shared/src/main/scala/org/http4s/ember/core/ChunkedEncoding.scala

```diff
@@ -140,20 +142,37 @@ private[ember] object ChunkedEncoding {
     }
 
   private val lastChunk: Chunk[Byte] =
-    Chunk.byteVector((ByteVector('0') ++ crlf ++ crlf).compact)
+    Chunk.byteVector((ByteVector('0') ++ crlf).compact)
 
   /** Encodes chunk of bytes to http chunked encoding.
     */
-  def encode[F[_]]: Pipe[F, Byte, Byte] = {
+  def encode[F[_]: Functor](trailers: F[Headers]): Pipe[F, Byte, Byte] = {
     def encodeChunk(bv: ByteVector): Chunk[Byte] =
       if (bv.isEmpty) Chunk.empty
       else
         Chunk.byteVector(
           ByteVector.view(bv.size.toHexString.toUpperCase.getBytes) ++ crlf ++ bv ++ crlf
         )
-    _.mapChunks { ch =>
-      encodeChunk(ch.toByteVector)
-    } ++ Stream.chunk(lastChunk)
+    _.mapChunks(ch => encodeChunk(ch.toByteVector)) ++
+      Stream.chunk(lastChunk) ++
+      Stream.evalUnChunk(trailers.map(encodeTrailers(_))) ++
+      Stream.chunk(Chunk.byteVector(crlf))
+  }
+
+  private def encodeTrailers(trailers: Headers): Chunk[Byte] = {
+    val stringBuilder = new StringBuilder()
```

Comment:
As a small suggestion: a StringBuilder starts with a capacity of 16 characters, but it duplicates every time. https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html#StringBuilder()
https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/java/lang/AbstractStringBuilder.java

Perhaps it could be more efficient to first calculate the total size of the trailers names, and values, with the colons, and initialise the StringBuilder to that size.

It could also help to keep the same instance of `StringBuilder`, with its underlying array, from use to use.

Reply:
@diesalbla I agree. We encode headers in at least three places, ideally these could share a single, optimized method.

> Perhaps it could be more efficient to first calculate the total size of the trailers names, and values, with the colons, and initialise the StringBuilder to that size.

In that case, I think if we could work directly with a `Array[Char]` instead of a `StringBuilder`. See my comment in  https://github.com/http4s/http4s/pull/6753#issuecomment-1279670163.



> It could also help to keep the same instance of `StringBuilder`, with its underlying array, from use to use.

I thought about this as well, but this would have problems with thread-safety.